src/controller/eme-controller.ts

```diff
@@ -522,10 +522,32 @@ class EMEController extends EventHandler {
   }
 
   onMediaDetached () {
-    if (this._media) {
-      this._media.removeEventListener('encrypted', this._onMediaEncrypted);
+    const media = this._media;
+    if (!media) return;
+
+    media.removeEventListener('encrypted', this._onMediaEncrypted);
+
+    // Close all sessions and remove media keys from the video element.
+    Promise.all(this._mediaKeysList.map((mediaKeysListItem) => {
```

Comment:
Removing the `this._media` and `this._mediaKeysList` references should be done synchronously.

```suggestion
    const media = this._media;
    const mediaKeysList = this._mediaKeysList;
    if (!media) {
      return;
    }
    media.removeEventListener('encrypted', this._onMediaEncrypted);
    this._media = null;
    this._mediaKeysList = [];
 
    // Close all sessions and remove media keys from the video element.
    Promise.all(mediaKeysList.map((mediaKeysListItem) => {
```

Obviously `mediaKeysListItem.mediaKeysSession.close ` and `media.setMediaKeys` calls need to be async as you have them now, but when this is happening you cannot attach the same media element again. How can we expose this promise so that you can wait for detach to complete?

IMO, this shouldn't be a `onMediaDetached` event handler. `api.detachMedia` should call `detachMedia` on controllers that implement and return `Promise.all` of all returned promises. You don't have to do that now, but I'd like your opinion on this. For v1 we should make api methods that need to be async like this example return promises.

Reply:
@springuper would you mind trying the suggested change above and letting me know if we can include these changes safely? Thanks!